### PR TITLE
New garbage collector implementation

### PR DESCRIPTION
**Pull request by sebastiencs**
- The garbage collector now keeps 3 block levels instead of 7 cycles
- `jemallocator` (unmaintained) is replaced with `tikv-jemallocator`
- Include the block hash in the error message when the block application fail @binier

### CODE REVIEW

File: tezos/context/src/gc/jemalloc.rs [hunk 1]
```diff
@@ -0,0 +1,80 @@
+use std::time::Duration;
```
Comment (tizoc) [inline tezos/context/src/gc/jemalloc.rs:1]: Copyright

**Review by tizoc — COMMENTED** [review]
@sebastiencs I need to do another pass, but something that would help a lot is an description of the new data structures, and how they improve things.

File: tezos/context/src/kv_store/in_memory.rs [hunk 2]
```diff
@@ -96,97 +107,173 @@ impl HashValueStore {
...
-        })
+        Ok(vacant)
     }
 
     fn get_free_id(&mut self) -> Option<HashId> {
         self.free_ids.as_mut()?.pop().ok()
     }
 
-    pub(crate) fn insert_value_at(
+    pub(crate) fn insert_object_at(
```
Comment (tizoc) [inline tezos/context/src/kv_store/in_memory.rs:141]: same here and everywhere the change from `value -> object` has been made

File: tezos/context/src/kv_store/in_memory.rs [hunk 3]
```diff
@@ -372,136 +457,263 @@ impl InMemory {
...
         })
     }
 
+    /// Reload context from disk
+    fn reload_database(&mut self) -> Result<(), ReloadError> {
+        debug_jemalloc();
+
+        let (tree, parent_hash, commit) = {
+            let mut ondisk = Persistent::try_new(PersistentConfiguration {
+                db_path: Some("/tmp/tezedge/context".to_string()),
```
Comment (tizoc) [inline tezos/context/src/kv_store/in_memory.rs:492]: @sebastiencs there is a hardcoded path here

File: tezos/context/src/working_tree/shape.rs [hunk 4]
```diff
@@ -129,9 +126,9 @@ impl DirectoryShapes {
     pub fn new() -> Self {
         Self {
             hash_to_strings: Default::default(),
-            id_to_hash: IndexMap::with_chunk_capacity(1_024 * 100),
+            id_to_hash: IndexMap::default(),
             temp: Vec::with_capacity(256),
-            shapes: ChunkedVec::with_chunk_capacity(64 * 1024 * 1024), // 64 MB
+            shapes: ChunkedVec::default(), // 64 MB
```
Comment (tizoc) [inline tezos/context/src/working_tree/shape.rs:131]: Is this size still correct?

File: tezos/context/src/working_tree/storage.rs [hunk 5]
```diff
@@ -1161,18 +1163,18 @@ const DEFAULT_THIN_POINTERS_CAPACITY: usize = 128 * 1024;
 impl Storage {
     pub fn new() -> Self {
         Self {
-            directories: ChunkedVec::with_chunk_capacity(DEFAULT_DIRECTORIES_CAPACITY), // ~4MB
-            temp_dir: Vec::with_capacity(256),                                          // 2KB
+            directories: ChunkedVec::default(), // ~4MB
```
Comment (tizoc) [inline tezos/context/src/working_tree/storage.rs:1166]: Same with all these sizes, not sure if still correct

**Review by tizoc — APPROVED** [review]
@sebastiencs overall looks good, but please check the comments I left (specially the one about the hardcoded path when restoring the in-memory context from disk)

File: tezos/context/src/kv_store/in_memory.rs [hunk 6]
```diff
@@ -69,25 +83,22 @@ impl HashValueStore {
         &self,
         strings_total_bytes: usize,
         shapes_total_bytes: usize,
-        commit_index_total_bytes: usize,
         nshapes: usize,
     ) -> RepositoryMemoryUsage {
         let values_bytes = self.values_bytes;
-        let values_capacity = self.values.capacity();
+        let values_capacity = self.objects.capacity();
```
[the remark this answers is not in the document]
Comment (sebastiencs) [reply tezos/context/src/kv_store/in_memory.rs:89]: Yes those are serialized objects: commits, directories and blobs

**Pull request merged**